### PR TITLE
Add new qaqc_outlier.sql for report generation

### PR DESCRIPTION
**Pull request by Oysters1874**
Add one additional qaqc_outlier.sql for outlier report generation. 

There are several questions that I would like to address. 

1. Currently, we are just generating new csv file for the outlier. I am using a similar format as qaqc_expected did, which does not take condo and mapped option into consideration. I am not sure whether this should be the case for the outlier report. It seems like we are just sorting out weird records, and it shouldn't be affected by the condo or mapped option.

### CODE REVIEW

**Comment by SashaWeinstein** [comment]
Have been looking through this PR this morning, I'm quite impressed so far. The JSON it produces has a bunch of information organized in a pretty intuitive way

File: pluto_build/sql/qaqc_outlier.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
+-- -v VERSION=$VERSION
+DROP TABLE IF EXISTS qaqc_outlier;
```
Comment (SashaWeinstein) [inline pluto_build/sql/qaqc_outlier.sql:2]: This should come out as going forward the QAQC should append new records

File: pluto_build/sql/qaqc_outlier.sql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    select jsonb_agg(json_build_object('pair',m.pair, 'bbl', m.bbl,
+                                       'building_area_current',m.building_area_current,
+                                       'building_area_previous',m.building_area_previous)) as values, 
+                                       'building_area_increase' as field
+    from (SELECT :'VERSION'||' - '||:'VERSION_PREV' as pair, 
+                a.bbl, a.lotarea as building_area_current, b.lotarea as building_area_previous
+          FROM archive_pluto a, previous_pluto b
+          WHERE a.bbl::FLOAT = b.bbl::FLOAT AND a.lotarea::FLOAT > 2*b.lotarea::FLOAT
+          AND b.lotarea::FLOAT != 0 ) m
+      )t);
```
Comment (SashaWeinstein) [inline pluto_build/sql/qaqc_outlier.sql:38]: The results of this query look good when I pulled them into python, still reviewing

**Review by SashaWeinstein — COMMENTED** [review]
I'm still reviewing, but wanted to give feedback that we need to pull down the existing table with `import_qaqc`, that code should be added to `05_qaqc.sh`. We can generate the table for the first time with this line commented out and move it to main manually

**Comment by SashaWeinstein** [comment]
Also Jingyi when you get a chance you should run the export step to make sure that it works, and we can review the files in the `371-create-new-qaqc-outlier-report-for-pluto` folder

File: pluto_build/sql/qaqc_outlier.sql [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	select jsonb_agg(json_build_object('bbl', a.bbl, 'unitsres',a.unitsres, 'resarea', a.resarea,
+                              'res_unit_ratio', a.res_unit_ratio)) as values, 'unitsres_resarea' as field
+	from (SELECT DISTINCT bbl, unitsres, resarea, resarea::INT/unitsres::INT as res_unit_ratio
+          FROM archive_pluto
+          WHERE unitsres::INT>50 and resarea::INT/unitsres::INT < 300
+          AND resarea::INT != 0) a
+	union
+	select jsonb_agg(json_build_object('bbl', a.bbl, 'bldgarea',a.bldgarea, 'lotarea', a.lotarea,
+                              'numfloors', a.numfloors, 'blog_lot_ratio', a.bldg_lot_ratio)) as values, 'lotarea_numfloor' as field
+	from (SELECT bbl, bldgarea, lotarea, numfloors, bldgarea::FLOAT/lotarea::FLOAT as bldg_lot_ratio
```
Comment (SashaWeinstein) [inline pluto_build/sql/qaqc_outlier.sql:24]: When I run this on my machine I get 1914 bbl's that match this criteria, but 1510 (~79%) have numfloors=0. It seems like these records have a separate problem that should go in their own report and not be lumped in with the lotarea/numfloor report. Curious what @AmandaDoyle thinks about this too

**Comment by SashaWeinstein** [comment]
The last query looks good, results looked good when I ran it on my machine.
This is just an idea, but is it possible to save the outlier column as a dictionary that maps the field (`unitsres_resarea`, `lotarea_numfloor`, `building_area_increase`) to a list of values instead of a list of 3 dictionaries? Would make unpacking the data on the app side marginally easier. I haven't worked with the `jsonb_agg` function much myself so I have no clue how hard this would be to implement

File: pluto_build/sql/qaqc_outlier.sql (hunk 3 above)
Comment (Oysters1874) [reply pluto_build/sql/qaqc_outlier.sql:24]: Yea, I agree with you. We should set an extra condition that numfloors doesn't equal to 0.

File: pluto_build/sql/qaqc_outlier.sql [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+DELETE FROM qaqc_outlier
+WHERE v = :'VERSION';
+
+INSERT INTO qaqc_outlier (
+select :'VERSION' as v, 	    
+        jsonb_agg(t) as outlier
+from (
+	select jsonb_agg(json_build_object('bbl', a.bbl, 'unitsres',a.unitsres, 'resarea', a.resarea,
+                              'res_unit_ratio', a.res_unit_ratio)) as values, 'unitsres_resarea' as field
+	from (SELECT DISTINCT bbl, unitsres, resarea, resarea::INT/unitsres::INT as res_unit_ratio
```
[the remark this answers is not in the document]
Comment (Oysters1874) [reply pluto_build/sql/qaqc_outlier.sql:17]: I am not sure which part of the code actually truncates the value, but I think for this part, just using float will be fine?

[the remark this answers is not in the document]
Comment (Oysters1874) [reply pluto_build/sql/qaqc_outlier.sql]: oh ok

**Comment by Oysters1874** [comment]
> 

right, before I haven't taken the condo and mapped thing into account. Will modify this part after making the code change.

**Comment by Oysters1874** [comment]
> The last query looks good, results looked good when I ran it on my machine. This is just an idea, but is it possible to save the outlier column as a dictionary that maps the field (`unitsres_resarea`, `lotarea_numfloor`, `building_area_increase`) to a list of values instead of a list of 3 dictionaries? Would make unpacking the data on the app side marginally easier. I haven't worked with the `jsonb_agg` function much myself so I have no clue how hard this would be to implement

Okay, I am looking at it. For each outlier type among the three field (`unitsres_resarea`, `lotarea_numfloor`, `building_area_increase`), there are multiple columns recording different fields of pluto. Since we want to display all three sections in table format, and finally turn each dictionary into dataframe, I guess probably mapping them to dictionary works better? I am not quite sure about how to map these fields to list of values, but I will try to make the code simpler.

File: pluto_build/sql/qaqc_outlier.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+INSERT INTO qaqc_outlier (
+select :'VERSION' as v, 	 
+       :CONDO as condo,
+       :MAPPED as mapped,
+       jsonb_agg(t) as outlier
+from (
+	select jsonb_agg(json_build_object('bbl', tmp.bbl, 'unitsres',tmp.unitsres, 'resarea', tmp.resarea,
+                              'res_unit_ratio', tmp.res_unit_ratio)) as values, 'unitsres_resarea' as field
+	from (SELECT bbl, unitsres, resarea, resarea::INT/unitsres::INT as res_unit_ratio
+          FROM (SELECT a.*
```
Comment (Oysters1874) [inline pluto_build/sql/qaqc_outlier.sql:24]: I am adding another subquery for the condo and mapped condition here, as I am not quite sure about how to really concatenate the input argument with other conditions in the where clause. I have tried several methods but seems they don't work properly. 

If anyone can come up with a nicer solution to make the code simpler, please point out for me. I feel the code is a bit too long right now, which is not really necessary.

**Review by SashaWeinstein — COMMENTED** [review]
I will test out this new field now. I think the two outstanding things before I approve are:
1) Change the cast in unitres vs resarea from INT to float so it doesn't truncate the ratio
2) Check the mapped/condo conditions for lot floor area and building area increase. This check doesn't have to as comprehensive as what I did [here](https://github.com/NYCPlanning/db-pluto/tree/Check-Outlier-Report) but we should make sure the mapped/condo condition worked for both

File: pluto_build/sql/qaqc_outlier.sql (hunk 1 above)
Comment (SashaWeinstein) [reply pluto_build/sql/qaqc_outlier.sql:2]: Missed this update in my comment yesterday, but this also needs to come out before merging

**Comment by Oysters1874** [comment]
> I will test out this new field now. I think the two outstanding things before I approve are:
> 
> 1. Change the cast in unitres vs resarea from INT to float so it doesn't truncate the ratio
> 2. Check the mapped/condo conditions for lot floor area and building area increase. This check doesn't have to as comprehensive as what I did [here](https://github.com/NYCPlanning/db-pluto/tree/Check-Outlier-Report) but we should make sure the mapped/condo condition worked for both

I have checked for both lot floor area and building area increase using all combinations of mapped and condo conditions, following the same way as you did. It can be confirmed now the mapped/condo filter works well for all metrics. But I am not sure whether I need to upload my check anywhere. Because I have trouble connecting to postgresql, I read from my local csv file and compare with query results in pgadmin.

File: pluto_build/sql/qaqc_outlier.sql (hunk 1 above)
Comment (Oysters1874) [reply pluto_build/sql/qaqc_outlier.sql]: Yea, thanks for pointing this out. I forgot to take it out before and just deleted it in the latest commit.

File: pluto_build/sql/qaqc_outlier.sql [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          AND lotarea::FLOAT != 0 AND numfloors !=0)tmp
+    union 
+    select jsonb_agg(json_build_object('pair',m.pair, 'bbl', m.bbl,
+                                       'building_area_current',m.building_area_current,
+                                       'building_area_previous',m.building_area_previous,
+                                       'percent_change',m.percent_change)) as values, 
+                                       'building_area_increase' as field
+    from (SELECT :'VERSION'||' - '||:'VERSION_PREV' as pair,
+                a.bbl, a.lotarea::FLOAT as building_area_current, b.lotarea::FLOAT as building_area_previous, 
+                (a.lotarea::FLOAT-b.lotarea::FLOAT)/b.lotarea::FLOAT as percent_change
```
Comment (SashaWeinstein) [inline pluto_build/sql/qaqc_outlier.sql:45]: This is almost there, need to multiple by 100 to get into units of percentage or change field name.
I also wanted to bring up my own confusion with the term "percent change". I looked at the numbers and the first row in the table is bbl `1002410018`. The current area is `326512`, and the previous `137575`. The current is 2.373 times the previous. I see that you are using [this formula for percent change](https://en.wikipedia.org/wiki/Relative_change_and_difference#Percentage_change) which gets the answer of 137.3% percent change. 
This seems to be technically correct, but I want to flag that to me a 137.3% "percent change" means going from 100 to 137.3. Whereas using the formula on wikipedia, it means adding 137.3% of the original total to go to 237.3. I wanted to bring @NYCPlanning/data-engineering's attention this question so we can get on the same page about what a "percent change" means (we should probably try to stick to the correct definition like Jingyi does here).

**Review by SashaWeinstein — COMMENTED** [review]
Really close, one small change to make

File: pluto_build/sql/qaqc_outlier.sql (hunk 6 above)
Comment (Oysters1874) [reply pluto_build/sql/qaqc_outlier.sql:45]: okay, this makes sense. As we are sorting out all BBLs that increased more than doubled than previous version, which means the percent change (in my definition or wiki) should be greater than 100%.

**Comment by SashaWeinstein** [comment]
I think it's ok if you don't upload your checks anywhere! Just as long as you do the check

File: pluto_build/05_qaqc.sh [hunk 7]
```diff
@@ -10,6 +10,7 @@ import_qaqc qaqc_expected main &
 import_qaqc qaqc_aggregate main &
 import_qaqc qaqc_mismatch main &
 import_qaqc qaqc_null main &
+#import_qaqc qaqc_outlier main &
```
Comment (SashaWeinstein) [inline pluto_build/05_qaqc.sh:13]: Ok sorry one last thing I noticed, uncomment this line. It will fail the first time but work after that

**Review by SashaWeinstein — APPROVED** [review]
Ready to be merged in! Great job

**Pull request merged**